pkg/blobinfocache/boltdb/boltdb.go

```diff
@@ -348,38 +390,37 @@ func (bdc *cache) candidateLocations(transport types.ImageTransport, scope types
 		// and we don't want to fail just because of that
 		compressionBucket := tx.Bucket(digestCompressorBucket)
 
-		res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, primaryDigest, requireCompressionInfo)
+		res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, primaryDigest, v2Output)
 		if canSubstitute {
-			if uncompressedDigestValue = bdc.uncompressedDigest(tx, primaryDigest); uncompressedDigestValue != "" {
-				b := tx.Bucket(digestByUncompressedBucket)
-				if b != nil {
-					b = b.Bucket([]byte(uncompressedDigestValue.String()))
-					if b != nil {
-						if err := b.ForEach(func(k, _ []byte) error {
-							d, err := digest.Parse(string(k))
-							if err != nil {
-								return err
-							}
-							if d != primaryDigest && d != uncompressedDigestValue {
-								res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, d, requireCompressionInfo)
-							}
-							return nil
-						}); err != nil {
-							return err
-						}
-					}
-				}
-				if uncompressedDigestValue != primaryDigest {
-					res = bdc.appendReplacementCandidates(res, scopeBucket, compressionBucket, uncompressedDigestValue, requireCompressionInfo)
-				}
+			res, err = getCandidatesForUncompressedDigest(tx, res, scopeBucket, compressionBucket, primaryDigest, v2Output)
+			if err != nil {
+				return err
 			}
 		}
 		return nil
 	}); err != nil { // Including os.IsNotExist(err)
 		return []blobinfocache.BICReplacementCandidate2{} // FIXME? Log err (but throttle the log volume on repeated accesses)?
 	}
+	// priortize the result
+	resultPrioritized := prioritize.DestructivelyPrioritizeReplacementCandidates(res, primaryDigest, uncompressedDigestValue)
 
-	return prioritize.DestructivelyPrioritizeReplacementCandidates(res, primaryDigest, uncompressedDigestValue)
+	// Reprocess all blobs and append them to resultPrioritized in lower priority
+	if err := bdc.view(func(tx *bolt.Tx) error {
+		compressionBucket := tx.Bucket(digestCompressorBucket)
+		resAllBlobs, err = getCandidatesForUncompressedDigest(tx, resAllBlobs, nil, compressionBucket, primaryDigest, v2Output)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil { // Including os.IsNotExist(err)
+		return []blobinfocache.BICReplacementCandidate2{} // FIXME? Log err (but throttle the log volume on repeated accesses)?
+	}
+	resultPrioritizedAllBlobs := prioritize.DestructivelyPrioritizeReplacementCandidates(resAllBlobs, primaryDigest, uncompressedDigestValue)
+	// Append `resultPrioritizedAllBlobs` after blobs
+	// which are generated from bucket with provided `scope`
+	// as a result `resultPrioritizedAllBlobs` will always
+	// get lower priority while processing candidates.
+	return append(resultPrioritized, resultPrioritizedAllBlobs...)
```

Comment:
The relative priority of known-location vs. unknown-location entries is not backend-dependent, and would be better handled in `prioritize` — along with the smaller limits for unknown-location entries.

Same in the memory backend.

Reply:
Yes moved to backend and now `prioritize` automatically takes care of this.